core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/stacks/loadbalancer/LoadBalancerToLoadBalancerResponseConverter.java

```diff
@@ -0,0 +1,124 @@
+
+@Component
+public class LoadBalancerToLoadBalancerResponseConverter extends AbstractConversionServiceAwareConverter<LoadBalancer, LoadBalancerResponse> {
+
+    @Inject
+    private TargetGroupPersistenceService targetGroupService;
+
+    @Inject
+    private InstanceGroupService instanceGroupService;
+
+    @Inject
+    private InstanceMetaDataService instanceMetaDataService;
+
+    @Inject
+    private LoadBalancerConfigService loadBalancerConfigService;
+
+    @Override
+    public LoadBalancerResponse convert(LoadBalancer source) {
+        LoadBalancerResponse response = new LoadBalancerResponse();
+        response.setIp(source.getIp());
+        response.setCloudDns(source.getDns());
+        response.setFqdn(source.getFqdn());
+        response.setType(source.getType());
+        response.setTargets(convertTargetGroups(targetGroupService.findByLoadBalancerId(source.getId())));
+        if (source.getProviderConfig() != null) {
+            response.setAwsResourceId(convertAwsLoadBalancer(source.getProviderConfig().getAwsConfig()));
+        }
+
+        return response;
+    }
+
+    private AwsLoadBalancerResponse convertAwsLoadBalancer(AwsLoadBalancerConfigDb awsMetadata) {
+        if (awsMetadata != null) {
+            AwsLoadBalancerResponse awsSettings = new AwsLoadBalancerResponse();
+            awsSettings.setArn(awsMetadata.getArn());
+            return awsSettings;
+        }
+        return null;
+    }
+
+    public List<TargetGroupResponse> convertTargetGroups(Set<TargetGroup> targetGroups) {
+        List<TargetGroupResponse> allResponses = new ArrayList<>();
+        for (TargetGroup targetGroup : targetGroups) {
+            allResponses.addAll(convertTargetGroup(targetGroup));
+        }
+        return allResponses;
+    }
+
+    public List<TargetGroupResponse> convertTargetGroup(TargetGroup targetGroup) {
+        Set<InstanceGroup> instanceGroups = instanceGroupService.findByTargetGroupId(targetGroup.getId());
+        Set<String> instanceIds = getInstanceMetadataForGroups(instanceGroups).stream()
+            .map(InstanceMetaData::getInstanceId)
+            .collect(Collectors.toSet());
+        List<TargetGroupResponse> responses = new ArrayList<>();
+        TargetGroupConfigDbWrapper targetGroupConfig = targetGroup.getProviderConfig();
+        Set<TargetGroupPortPair> portPairs = loadBalancerConfigService.getTargetGroupPortPairs(targetGroup);
+
+        for (TargetGroupPortPair portPair : portPairs) {
+            TargetGroupResponse response = new TargetGroupResponse();
+            response.setPort(portPair.getTrafficPort());
+            response.setTargetInstances(instanceIds);
+            if (targetGroupConfig != null) {
+                response.setAwsResourceIds(convertAwsTargetGroup(targetGroupConfig.getAwsConfig(), portPair.getTrafficPort()));
+            }
+            responses.add(response);
+        }
+
+        return responses;
```

Comment:
this could be a bit improved I think, like:
```
        return portPairs.stream()
                .map(portPair -> mapPortPairToTargetGroup(instanceIds, targetGroupConfig, portPair))
                .collect(Collectors.toList());
    }

    private TargetGroupResponse mapPortPairToTargetGroup(Set<String> instanceIds, TargetGroupConfigDbWrapper targetGroupConfig, TargetGroupPortPair portPair) {
        TargetGroupResponse response = new TargetGroupResponse();
        response.setPort(portPair.getTrafficPort());
        response.setTargetInstances(instanceIds);
        if (targetGroupConfig != null) {
            response.setAwsResourceIds(convertAwsTargetGroup(targetGroupConfig.getAwsConfig(), portPair.getTrafficPort()));
        }
        return response;
    }

Reply:
Done. Thanks for the code.